File: object-oriented-programming/overview/python/pet_class.py
# Raspberry Pi Foundation
# Developed as part of Ada Computer Science
# 
# Usage licensed under CC BY-NC-SA 4.0

import random

class Pet:

    # Constructor method
    def __init__(self, given_name, given_type, given_colour):
        self.name = given_name
        self.pet_type = given_type
        self.colour = given_colour
        self.sleeping = False

        # Randomly choose a starting mood
        all_moods = ["playful", "hungry", "tired"]
        self.mood = random.choice(all_moods)

    def get_name(self):
        return self.name

    def set_name(self, new_name):
        self.name = new_name

    def get_pet_type(self):
        return self.pet_type

    def set_pet_type(self, new_type):
        self.pet_type = new_type

    def get_colour(self):
        return self.colour

    def set_colour(self, new_colour):
        self.colour = new_colour

    def get_mood(self):
        return self.mood

    def is_sleeping(self):
        return self.sleeping

    def describe(self):
        print(f"I am a {self.mood}, {self.colour} {self.pet_type} called {self.name}")

    def play(self):
        """A method for trying to play with the pet"""
        if self.sleeping == True:
            print("Zzzzzzz. I am sleeping")
        elif self.mood == "hungry":
            print("I am too hungry to play")
        elif self.mood == "tired":
            print("I am too tired to play")
        else:
            print("This is fun, I love playing")

            # Randomly choose a new mood
            all_moods = ["playful", "hungry", "tired"]
            self.mood = random.choice(all_moods)

    def feed(self):
        """A method for trying to feed the pet"""
        if self.sleeping == True:
            print("Zzzzzzz. I am sleeping")
        elif self.mood == "tired":
            print("I am too sleepy to eat anything now")
        elif self.mood == "playful":
            print("I am not hungry - I want to play")
        else:
            print("Yum - yum - that tastes great")
            self.mood = "playful"

    def sleep(self):
        """A method for trying to put the pet to bed"""
        if self.mood == "playful":
            print("I am too playful to sleep")
        elif self.mood == "hungry":
            print("I need something to eat before I can go to sleep")
        else:
            print("Zzzzzzzz")
            self.sleeping = True

    def wake(self):
        """A method for trying to wake the pet up"""
        if self.sleeping == False:
           print("I wasn't even asleep!")
        else:
            print("I am awake and ready to play!")
            self.sleeping = False
            self.mood = "playful"

File: object-oriented-programming/overview/java/pet_class.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

import java.util.Random;

class Pet {
    private String name;
    private String petType;
    private String colour;
    private String mood;
    private boolean sleeping;

    // Constructor method
    public Pet(String givenName, String givenType, String givenColour) {
        name = givenName;
        petType = givenType;
        colour = givenColour;
        sleeping = false;

        // Randomly choose a starting mood
        String[] allMoods = { "playful", "hungry", "tired" };
        Random rnd = new Random();
        int index = rnd.nextInt(allMoods.length);
        mood = allMoods[index];
    }

    public String getName() {
        return name;
    }

    public void setName(String newName) {
        name = newName;
    }

    public String getPetType() {
        return petType;
    }

    public void setPetType(String newType) {
        petType = newType;
    }

    public String getColour() {
        return colour;
    }

    public void setColour(String newColour) {
        colour = newColour;
    }

    public String getMood() {
        return mood;
    }

    public boolean isSleeping() {
        return sleeping;
    }

    public void describe() {
        System.out.println("I am a " + mood + ", " + colour + " " + petType + " called " + name);
    }


    // A method for trying to play with the pet
    public void play() {
        if (sleeping == true) {
            System.out.println("Zzzzzzz. I am sleeping");
        }
        else if ("hungry".equals(mood)) {
            System.out.println("I am too hungry to play");
        }
        else if ("tired".equals(mood)) {
            System.out.println("I am too tired to play");
        }
        else {
            System.out.println("This is fun, I love playing");

            // Randomly choose a new mood
            String[] allMoods = { "playful", "hungry", "tired" };
            Random rnd = new Random();
            int index = rnd.nextInt(allMoods.length);
            mood = allMoods[index];
        }
    }


    // A method for trying to feed the pet
    public void feed() {
        if (sleeping == true) {
            System.out.println("Zzzzzzz. I am sleeping");
        }
        else if ("tired".equals(mood)) {
            System.out.println("I am too sleepy to eat anything now");
        }
        else if ("playful".equals(mood)) {
            System.out.println("I am not hungry - I want to play");
        }
        else {
            System.out.println("Yum - yum - that tastes great");
            mood = "playful";
        }
    }


    // A method for trying to put the pet to bed
    public void sleep() {
        if ("playful".equals(mood)) {
            System.out.println("I am too playful to sleep");
        }
        else if ("hungry".equals(mood)) {
            System.out.println("I need something to eat before I can go to sleep");
        }
        else {
            System.out.println("Zzzzzzzz");
            sleeping = true;
        }
    }


    // A method for trying to wake the pet up
    public void wake() {
        if (sleeping == false) {
            System.out.println("I wasn't even asleep!");
        }
        else {
            System.out.println("I am awake and ready to play!");
            sleeping = false;
            mood = "playful";
        }
    }


}